File: frontend/src/features/maps/model/routePresentation.ts
import type {
  MakersMapPlace,
  MakersRouteLeg,
  MakersRouteMode,
  MakersRoutePlan,
  MakersRouteSection,
  MakersRouteSectionMode,
} from './types';

export type RouteZoomLevel = 'overview' | 'legs' | 'sections';

export interface RouteSectionStep {
  leg: MakersRouteLeg;
  legIndex: number;
  section: MakersRouteSection;
  sectionIndex: number;
}

export interface RouteSectionEndpoints {
  from: string;
  to: string;
}

export interface RouteMapPoint {
  latitude: number;
  longitude: number;
}

export const ROUTE_MODE_COLORS: Record<MakersRouteSectionMode, string> = {
  driving: '#e5484d',
  transit: '#2f6fed',
  bus: '#2f6fed',
  rail: '#7c3aed',
  walking: '#2e9d67',
  bicycling: '#ed8b2c',
};

export function routeZoomLevel(zoom: number): RouteZoomLevel {
  if (zoom <= 8) return 'overview';
  if (zoom <= 12) return 'legs';
  return 'sections';
}

function fallbackSection(
  mode: MakersRouteMode,
  path: MakersRoutePlan['path'],
  distance: number,
  duration: number,
): MakersRouteSection {
  return {
    mode,
    path,
    distance_meters: distance,
    duration_seconds: duration,
  };
}

export function routeLegs(route: MakersRoutePlan): MakersRouteLeg[] {
  if (route.legs?.length) return route.legs;
  if (route.places.length < 2) return [];
  return [{
    from: route.places[0],
    to: route.places[route.places.length - 1],
    mode: route.mode,
    path: route.path,
    sections: [fallbackSection(
      route.mode,
      route.path,
      route.distance_meters,
      route.duration_seconds,
    )],
    distance_meters: route.distance_meters,
    duration_seconds: route.duration_seconds,
  }];
}

export function routeLegScope(
  leg: MakersRouteLeg,
): 'intercity' | 'local' | 'unknown' {
  if (leg.scope) return leg.scope;
  const originCity = String(leg.from.city || '').trim().toLowerCase();
  const destinationCity = String(leg.to.city || '').trim().toLowerCase();
  if (!originCity || !destinationCity) return 'unknown';
  return originCity === destinationCity ? 'local' : 'intercity';
}

export function routeHasIntercityLeg(route: MakersRoutePlan): boolean {
  return routeLegs(route).some((leg) => routeLegScope(leg) === 'intercity');
}

export function visibleRouteSections(route: MakersRoutePlan): MakersRouteSection[] {
  return routeSectionSteps(route).map((step) => ({
    ...step.section,
    path: routeSectionPath(step),
  }));
}

export function routeSectionSteps(route: MakersRoutePlan): RouteSectionStep[] {
  return routeLegs(route).flatMap((leg, legIndex) => {
    const sections = leg.sections?.length
      ? leg.sections
      : [fallbackSection(leg.mode, leg.path, leg.distance_meters, leg.duration_seconds)];
    return sections.map((section, sectionIndex) => ({
      leg,
      legIndex,
      section,
      sectionIndex,
    }));
  });
}

/**
 * Derive a label only from provider-backed stop names around one section.
 * This avoids presenting an unlabelled transfer walk as the entire route leg.
 */
export function routeSectionEndpoints(
  steps: RouteSectionStep[],
  index: number,
): RouteSectionEndpoints | null {
  const current = steps[index];
  if (!current) return null;
  const explicitFrom = String(current.section.geton || '').trim();
  const explicitTo = String(current.section.getoff || '').trim();
  if (explicitFrom && explicitTo) return { from: explicitFrom, to: explicitTo };

  let from = String(current.leg.from.name || '').trim();
  for (let cursor = index - 1; cursor >= 0; cursor -= 1) {
    const previous = steps[cursor];
    if (previous.legIndex !== current.legIndex) break;
    const endpoint = String(
      previous.section.getoff || previous.section.geton || '',
    ).trim();
    if (endpoint) {
      from = endpoint;
      break;
    }
  }

  let to = String(current.leg.to.name || '').trim();
  for (let cursor = index + 1; cursor < steps.length; cursor += 1) {
    const next = steps[cursor];
    if (next.legIndex !== current.legIndex) break;
    const endpoint = String(next.section.geton || next.section.getoff || '').trim();
    if (endpoint) {
      to = endpoint;
      break;
    }
  }
  return from && to && from !== to ? { from, to } : null;
}

/**
 * Older provider snapshots may only contain a leg polyline. In that case,
 * divide it proportionally by the provider section distances so every client
 * can still focus the relevant section without inventing route semantics.
 */
export function routeSectionPath({ leg, section, sectionIndex }: RouteSectionStep) {
  if (section.path.length > 1 || leg.path.length < 2) return section.path;
  const sections = leg.sections?.length ? leg.sections : [section];
  const weights = sections.map((item) => (
    item.distance_meters > 0 ? item.distance_meters
      : item.duration_seconds > 0 ? item.duration_seconds
        : 1
  ));
  const total = weights.reduce((sum, weight) => sum + weight, 0);
  const before = weights.slice(0, sectionIndex).reduce((sum, weight) => sum + weight, 0);
  const after = before + (weights[sectionIndex] || 1);
  const lastPoint = leg.path.length - 1;
  const start = Math.max(0, Math.floor((before / total) * lastPoint));
  const end = Math.min(lastPoint, Math.ceil((after / total) * lastPoint));
  return leg.path.slice(start, Math.max(start + 2, end + 1));
}

function planarMetersSquared(a: RouteMapPoint, b: RouteMapPoint): number {
  const latitude = ((a.latitude + b.latitude) / 2) * (Math.PI / 180);
  const dx = (a.longitude - b.longitude) * 111_320 * Math.cos(latitude);
  const dy = (a.latitude - b.latitude) * 110_540;
  return (dx * dx) + (dy * dy);
}

function segmentDistanceSquaredMeters(
  point: RouteMapPoint,
  start: RouteMapPoint,
  end: RouteMapPoint,
): number {
  const latitude = point.latitude * (Math.PI / 180);
  const scaleX = 111_320 * Math.cos(latitude);
  const scaleY = 110_540;
  const ax = start.longitude * scaleX;
  const ay = start.latitude * scaleY;
  const bx = end.longitude * scaleX;
  const by = end.latitude * scaleY;
  const px = point.longitude * scaleX;
  const py = point.latitude * scaleY;
  const dx = bx - ax;
  const dy = by - ay;
  const length = (dx * dx) + (dy * dy);
  const ratio = length
    ? Math.max(0, Math.min(1, (((px - ax) * dx) + ((py - ay) * dy)) / length))
    : 0;
  const offsetX = px - (ax + (ratio * dx));
  const offsetY = py - (ay + (ratio * dy));
  return (offsetX * offsetX) + (offsetY * offsetY);
}

function simplifyPresentationPath(
  path: RouteMapPoint[],
  toleranceMeters: number,
): RouteMapPoint[] {
  if (path.length < 3) return path;
  const keep = new Set([0, path.length - 1]);
  const toleranceSquared = toleranceMeters * toleranceMeters;
  const pending: Array<[number, number]> = [[0, path.length - 1]];
  while (pending.length) {
    const [startIndex, endIndex] = pending.pop() as [number, number];
    let farthestIndex = -1;
    let farthestDistance = 0;
    for (let index = startIndex + 1; index < endIndex; index += 1) {
      const distance = segmentDistanceSquaredMeters(
        path[index], path[startIndex], path[endIndex],
      );
      if (distance > farthestDistance) {
        farthestDistance = distance;
        farthestIndex = index;
      }
    }
    if (farthestIndex < 0 || farthestDistance <= toleranceSquared) continue;
    keep.add(farthestIndex);
    pending.push([startIndex, farthestIndex], [farthestIndex, endIndex]);
  }
  return [...keep].sort((a, b) => a - b).map((index) => path[index]);
}

function dottedWalkingPaths(path: RouteMapPoint[]): RouteMapPoint[][] {
  const dots: RouteMapPoint[][] = [];
  const totalMeters = path.slice(0, -1).reduce(
    (sum, point, index) => sum + Math.sqrt(planarMetersSquared(point, path[index + 1])),
    0,
  );
  const cycleMeters = Math.max(46, totalMeters / 160);
  const dashMeters = cycleMeters * 0.61;
  const gapMeters = cycleMeters - dashMeters;
  let draw = true;
  let remaining = dashMeters;
  for (let index = 0; index < path.length - 1; index += 1) {
    const start = path[index];
    const end = path[index + 1];
    const distance = Math.sqrt(planarMetersSquared(start, end));
    if (!distance) continue;
    let offset = 0;
    while (offset < distance) {
      const length = Math.min(remaining, distance - offset);
      const fromRatio = offset / distance;
      const toRatio = (offset + length) / distance;
      if (draw && length > 0.5) {
        dots.push([
          {
            latitude: start.latitude + ((end.latitude - start.latitude) * fromRatio),
            longitude: start.longitude + ((end.longitude - start.longitude) * fromRatio),
          },
          {
            latitude: start.latitude + ((end.latitude - start.latitude) * toRatio),
            longitude: start.longitude + ((end.longitude - start.longitude) * toRatio),
          },
        ]);
      }
      offset += length;
      remaining -= length;
      if (remaining <= 0.5) {
        draw = !draw;
        remaining = draw ? dashMeters : gapMeters;
      }
    }
  }
  return dots.length ? dots : [path];
}

/**
 * Return render-only geometry for one provider section. Walking geometry is
 * lightly simplified and split into short round-capped strokes so transfer
 * details stay readable without altering the trusted route or its metrics.
 */
export function routeSectionDisplayPaths(step: RouteSectionStep): RouteMapPoint[][] {
  const path = routeSectionPath(step);
  if (step.section.mode !== 'walking' || path.length < 2) return [path];
  const tolerance = Math.max(
    6,
    Math.min(18, Number(step.section.distance_meters || 0) / 120),
  );
  return dottedWalkingPaths(simplifyPresentationPath(path, tolerance));
}

function projectedDistanceSquared(
  point: RouteMapPoint,
  path: RouteMapPoint[],
): number {
  if (!path.length) return Number.POSITIVE_INFINITY;
  const longitudeScale = Math.cos((point.latitude * Math.PI) / 180);
  const px = point.longitude * longitudeScale;
  const py = point.latitude;
  let closest = Number.POSITIVE_INFINITY;

  for (let index = 0; index < path.length; index += 1) {
    const start = path[index];
    const end = path[index + 1] || start;
    const ax = start.longitude * longitudeScale;
    const ay = start.latitude;
    const bx = end.longitude * longitudeScale;
    const by = end.latitude;
    const dx = bx - ax;
    const dy = by - ay;
    const lengthSquared = (dx * dx) + (dy * dy);
    const projection = lengthSquared
      ? Math.max(0, Math.min(1, (((px - ax) * dx) + ((py - ay) * dy)) / lengthSquared))
      : 0;
    const offsetX = px - (ax + (projection * dx));
    const offsetY = py - (ay + (projection * dy));
    closest = Math.min(closest, (offsetX * offsetX) + (offsetY * offsetY));
  }
  return closest;
}

/**
 * Resolve the route section nearest to the map's visual centre. This keeps the
 * UI driven by provider geometry and works for any city or transport mix.
 */
export function closestRouteSectionIndex(
  route: MakersRoutePlan,
  point: RouteMapPoint,
): number {
  const steps = routeSectionSteps(route);
  let closestIndex = -1;
  let closestDistance = Number.POSITIVE_INFINITY;
  steps.forEach((step, index) => {
    const path = routeSectionPath(step);
    const distance = projectedDistanceSquared(point, path);
    if (distance < closestDistance) {
      closestDistance = distance;
      closestIndex = index;
    }
  });
  return closestIndex;
}

export function routeCities(places: MakersMapPlace[]): Array<{
  city: string;
  place: MakersMapPlace;
}> {
  const seen = new Set<string>();
  const result: Array<{ city: string; place: MakersMapPlace }> = [];
  places.forEach((place) => {
    const city = String(place.city || '').trim();
    if (!city || seen.has(city)) return;
    seen.add(city);
    result.push({ city, place });
  });
  return result;
}

export function legModeSequence(leg: MakersRouteLeg): MakersRouteSectionMode[] {
  const result: MakersRouteSectionMode[] = [];
  (leg.sections || []).forEach((section) => {
    if (result[result.length - 1] !== section.mode) result.push(section.mode);
  });
  return result.length ? result : [leg.mode];
}

File: android/app/src/main/java/com/floris/android/core/maps/RoutePresentation.kt
package com.floris.android.core.maps

import com.floris.android.core.model.RouteLeg
import com.floris.android.core.model.RoutePlan
import com.floris.android.core.model.RoutePoint
import com.floris.android.core.model.RouteSection
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/** Render-only projection of provider-backed route data. No route is computed here. */
data class RouteSectionStep(
    val leg: RouteLeg,
    val legIndex: Int,
    val section: RouteSection,
    val sectionIndex: Int,
)

data class RouteSectionEndpoints(val from: String, val to: String)

fun routeLegs(route: RoutePlan): List<RouteLeg> {
    if (route.legs.isNotEmpty()) return route.legs
    val places = route.places.ifEmpty { route.ordered_stops }
    if (places.size < 2) return emptyList()
    return listOf(
        RouteLeg(
            from = places.first(),
            to = places.last(),
            mode = route.mode,
            path = route.path,
            sections = listOf(
                RouteSection(
                    mode = route.mode ?: "driving",
                    path = route.path,
                    distance_meters = route.distance_meters,
                    duration_seconds = route.duration_seconds,
                ),
            ),
            distance_meters = route.distance_meters,
            duration_seconds = route.duration_seconds,
        ),
    )
}

fun routeLegScope(leg: RouteLeg): String {
    leg.scope?.takeIf(String::isNotBlank)?.let { return it }
    val origin = leg.from?.city.orEmpty().trim().lowercase()
    val destination = leg.to?.city.orEmpty().trim().lowercase()
    return when {
        origin.isEmpty() || destination.isEmpty() -> "unknown"
        origin == destination -> "local"
        else -> "intercity"
    }
}

fun routeSectionSteps(route: RoutePlan): List<RouteSectionStep> =
    routeLegs(route).flatMapIndexed { legIndex, leg ->
        val sections = leg.sections.ifEmpty {
            listOf(
                RouteSection(
                    mode = leg.mode ?: "driving",
                    path = leg.path,
                    distance_meters = leg.distance_meters,
                    duration_seconds = leg.duration_seconds,
                ),
            )
        }
        sections.mapIndexed { sectionIndex, section ->
            RouteSectionStep(leg, legIndex, section, sectionIndex)
        }
    }

fun routeSectionEndpoints(steps: List<RouteSectionStep>, index: Int): RouteSectionEndpoints? {
    val current = steps.getOrNull(index) ?: return null
    val explicitFrom = current.section.geton.orEmpty().trim()
    val explicitTo = current.section.getoff.orEmpty().trim()
    if (explicitFrom.isNotEmpty() && explicitTo.isNotEmpty()) {
        return RouteSectionEndpoints(explicitFrom, explicitTo)
    }
    var from = current.leg.from?.name.orEmpty().trim()
    for (cursor in index - 1 downTo 0) {
        val previous = steps[cursor]
        if (previous.legIndex != current.legIndex) break
        val endpoint = (previous.section.getoff ?: previous.section.geton).orEmpty().trim()
        if (endpoint.isNotEmpty()) {
            from = endpoint
            break
        }
    }
    var to = current.leg.to?.name.orEmpty().trim()
    for (cursor in index + 1 until steps.size) {
        val next = steps[cursor]
        if (next.legIndex != current.legIndex) break
        val endpoint = (next.section.geton ?: next.section.getoff).orEmpty().trim()
        if (endpoint.isNotEmpty()) {
            to = endpoint
            break
        }
    }
    return if (from.isNotEmpty() && to.isNotEmpty() && from != to) {
        RouteSectionEndpoints(from, to)
    } else null
}

/** Split an older leg-only provider path proportionally across its real sections. */
fun routeSectionPath(step: RouteSectionStep): List<RoutePoint> {
    val leg = step.leg
    val section = step.section
    if (section.path.size > 1 || leg.path.size < 2) return section.path
    val sections = leg.sections.ifEmpty { listOf(section) }
    val weights = sections.map {
        when {
            it.distance_meters > 0 -> it.distance_meters
            it.duration_seconds > 0 -> it.duration_seconds
            else -> 1.0
        }
    }
    val total = weights.sum().takeIf { it > 0 } ?: 1.0
    val before = weights.take(step.sectionIndex).sum()
    val after = before + (weights.getOrNull(step.sectionIndex) ?: 1.0)
    val last = leg.path.lastIndex
    val start = ((before / total) * last).toInt().coerceIn(0, last)
    val end = kotlin.math.ceil((after / total) * last).toInt().coerceIn(0, last)
    return leg.path.subList(start, max(start + 2, end + 1).coerceAtMost(leg.path.size))
}

/** Render-only walking geometry: lightly simplify, then split into round-cap dots. */
fun routeSectionDisplayPaths(step: RouteSectionStep): List<List<RoutePoint>> {
    val path = routeSectionPath(step)
    if (step.section.mode != "walking" || path.size < 2) return listOf(path)
    val tolerance = max(6.0, min(18.0, step.section.distance_meters / 120.0))
    return dottedWalkingPaths(simplifyPath(path, tolerance))
}

private fun simplifyPath(path: List<RoutePoint>, toleranceMeters: Double): List<RoutePoint> {
    if (path.size < 3) return path
    val keep = mutableSetOf(0, path.lastIndex)
    val pending = ArrayDeque<Pair<Int, Int>>().apply { add(0 to path.lastIndex) }
    val toleranceSquared = toleranceMeters * toleranceMeters
    while (pending.isNotEmpty()) {
        val (start, end) = pending.removeLast()
        var farthest = -1
        var distance = 0.0
        for (index in start + 1 until end) {
            val candidate = segmentDistanceSquared(path[index], path[start], path[end])
            if (candidate > distance) {
                distance = candidate
                farthest = index
            }
        }
        if (farthest >= 0 && distance > toleranceSquared) {
            keep += farthest
            pending += start to farthest
            pending += farthest to end
        }
    }
    return keep.sorted().map(path::get)
}

private fun dottedWalkingPaths(path: List<RoutePoint>): List<List<RoutePoint>> {
    val totalMeters = path.zipWithNext().sumOf { (start, end) -> sqrt(metersSquared(start, end)) }
    val cycle = max(46.0, totalMeters / 160.0)
    val dash = cycle * 0.61
    val gap = cycle - dash
    var drawing = true
    var remaining = dash
    val result = mutableListOf<List<RoutePoint>>()
    path.zipWithNext().forEach { (start, end) ->
        val distance = sqrt(metersSquared(start, end))
        if (distance <= 0) return@forEach
        var offset = 0.0
        while (offset < distance) {
            val length = min(remaining, distance - offset)
            if (drawing && length > 0.5) {
                result += listOf(interpolate(start, end, offset / distance), interpolate(start, end, (offset + length) / distance))
            }
            offset += length
            remaining -= length
            if (remaining <= 0.5) {
                drawing = !drawing
                remaining = if (drawing) dash else gap
            }
        }
    }
    return result.ifEmpty { listOf(path) }
}

private fun interpolate(start: RoutePoint, end: RoutePoint, ratio: Double) = RoutePoint(
    latitude = start.latitude + (end.latitude - start.latitude) * ratio,
    longitude = start.longitude + (end.longitude - start.longitude) * ratio,
)

private fun metersSquared(a: RoutePoint, b: RoutePoint): Double {
    val latitude = ((a.latitude + b.latitude) / 2) * Math.PI / 180
    val dx = (a.longitude - b.longitude) * 111_320 * cos(latitude)
    val dy = (a.latitude - b.latitude) * 110_540
    return dx * dx + dy * dy
}

private fun segmentDistanceSquared(point: RoutePoint, start: RoutePoint, end: RoutePoint): Double {
    val latitude = point.latitude * Math.PI / 180
    val scaleX = 111_320 * cos(latitude)
    val scaleY = 110_540.0
    val ax = start.longitude * scaleX
    val ay = start.latitude * scaleY
    val bx = end.longitude * scaleX
    val by = end.latitude * scaleY
    val px = point.longitude * scaleX
    val py = point.latitude * scaleY
    val dx = bx - ax
    val dy = by - ay
    val length = dx * dx + dy * dy
    val ratio = if (length > 0) (((px - ax) * dx + (py - ay) * dy) / length).coerceIn(0.0, 1.0) else 0.0
    val ox = px - (ax + ratio * dx)
    val oy = py - (ay + ratio * dy)
    return ox * ox + oy * oy
}
